apps/admin/payments.py

```diff
@@ -390,41 +384,17 @@ def refund(payment_id):
 
     purchases_dict = {p.id: p for p in payment.purchases}
 
-    for f in form.purchases:
-        f._purchase = purchases_dict[f.purchase_id.data]
-        f.refund.label.text = "%s - %s" % (
-            f._purchase.id,
-            f._purchase.product.display_name,
-        )
-
-        if (
-            f._purchase.refund_id is None
-            and f._purchase.is_paid_for
-            and f._purchase.owner == payment.user
-        ):
-            # Purchase is owned by the user and not already refunded
-            f._disabled = False
-
-            if type(f._purchase) == AdmissionTicket and f._purchase.checked_in:
-                f.refund.data = False
-                f.refund.label.text += " (checked in)"
-        elif f._purchase.refund_id is not None:
-            f._disabled = True
-            f.refund.data = False
-            f.refund.label.text += " (refunded)"
-        else:
-            f._disabled = True
-            f.refund.data = False
-            f.refund.label.text += " (transferred)"
-
     if form.validate_on_submit():
         if form.refund.data or form.stripe_refund.data:
 
             payment.lock()
 
             purchases = [
-                f._purchase for f in form.purchases if f.refund.data and not f._disabled
+                purchases_dict[f.purchase_id.data]
+                for f in form.purchases
+                if f.refund.data and purchases_dict[f.purchase_id.data].is_refundable
```

Comment:
Should this be a method call?

```suggestion
                if f.refund.data and purchases_dict[f.purchase_id.data].is_refundable()
```

Reply:
Thanks - there are a few more of these though, all in `main.py` afaict.